lib/DxilValidation/DxilValidationUtils.h

```diff
@@ -0,0 +1,149 @@
+///////////////////////////////////////////////////////////////////////////////
+//                                                                           //
+// DxilValidationUttils.h                                                    //
+// Copyright (C) Microsoft Corporation. All rights reserved.                 //
+// This file is distributed under the University of Illinois Open Source     //
+// License. See LICENSE.TXT for details.                                     //
+//                                                                           //
+// This file provides utils for validating DXIL.                             //
+//                                                                           //
+///////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include "dxc/DXIL/DxilConstants.h"
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Analysis/CallGraph.h"
+#include "llvm/IR/DebugLoc.h"
+#include "llvm/IR/ModuleSlotTracker.h"
+
+#include <memory>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+using namespace llvm;
+
+namespace llvm {
+class Module;
+class Function;
+class DataLayout;
+class Metadata;
+class Value;
+class GlobalVariable;
+class Instruction;
+class Type;
+} // namespace llvm
+
+namespace hlsl {
+
+///////////////////////////////////////////////////////////////////////////////
+// Validation rules.
+#include "DxilValidation.inc"
+
+const char *GetValidationRuleText(ValidationRule value);
+
+class DxilEntryProps;
+class DxilModule;
+class DxilResourceBase;
+struct DxilResourceProperties;
+class DxilSignatureElement;
+
+// Save status like output write for entries.
+struct EntryStatus {
+  bool hasOutputPosition[DXIL::kNumOutputStreams];
+  unsigned OutputPositionMask[DXIL::kNumOutputStreams];
+  std::vector<unsigned> outputCols;
+  std::vector<unsigned> patchConstOrPrimCols;
+  bool m_bCoverageIn, m_bInnerCoverageIn;
+  bool hasViewID;
+  unsigned domainLocSize;
+  EntryStatus(DxilEntryProps &entryProps);
+};
+
+struct ValidationContext {
+  bool Failed = false;
+  Module &M;
+  Module *pDebugModule;
+  DxilModule &DxilMod;
+  const Type *HandleTy;
+  const DataLayout &DL;
+  DebugLoc LastDebugLocEmit;
+  ValidationRule LastRuleEmit;
+  std::unordered_set<Function *> entryFuncCallSet;
+  std::unordered_set<Function *> patchConstFuncCallSet;
+  std::unordered_map<unsigned, bool> UavCounterIncMap;
+  std::unordered_map<Value *, unsigned> HandleResIndexMap;
+  // TODO: save resource map for each createHandle/createHandleForLib.
+  std::unordered_map<Value *, DxilResourceProperties> ResPropMap;
```

Comment:
Looks like `DxilResourceProperties` is an incomplete type, never gets defined after its declaration on line 51.

Reply:
Fixed.